### PR TITLE
feat(workflow): add deploy preview links

### LINKED ISSUE #372: [FEATURE] Show Deploy Preview statuses and links in editorial workflow
**- Do you want to request a *feature* or report a *bug*?**

This is a feature that just crossed my mind.

**- What is the current behavior?**

Pages edited with the Editorial Workflow show up in the board, depending on their progress. If an editor wants to see the content rendered, they can see it with the builtin preview in the CMS. This is usually handy, but there is still a disconnection between what's rendered by the CMS and the actual page in its right domain.

If you use Netlify, you automatically get previews of the content edited in the right domain, the site deployed in our CDN. These previews are registered in GitHub's commit statuses.

So if an editor wants to check out the Deploy Preview, they need to know where the GitHub repository is, browse the list of pull requests, find the right one and click the link to open the Deploy Preview.

**- What is the expected behavior?**

GitHub provides an API to get information about those commit statuses:

https://developer.github.com/v3/repos/statuses/#get-the-combined-status-for-a-specific-ref

We could use it to check if there is a link to the Deploy Preview published there, then add it to the editorial workflow's card, and probably somewhere in the editor too. That way, if you want to check the edited content in the site, you only need to click a link that the CMS gives you.

As an initial implementation, I would not try to do anything super complicated, maybe not even for all cards, since all those calls to the GitHub API can be rate limited very fast. I would probably only check the content that's in "Waiting for Review", and only when the page is loaded. Polling data from GitHub's API can mark you as an abuser and you can lose access for up to 48 hours.

### PR DESCRIPTION
**Pull request by erquhart**
## Summary

Adds deploy preview links for unpublished entries in the editorial workflow. This PR adds the link in the editor toolbar for the current entry, if it's unpublished, and if a preview is available.

For now, this functionality is only provided in the GitHub backend, but it shouldn't take much to add it to the GitLab backend as well, or any other backend that supports the editorial workflow.

preview-link-check

preview-link-unpublished

preview-link-published

### How it works
**tl;dr** - it "just works" except for the "Link to the current entry" bit below.

The GitHub [Statuses API](https://developer.github.com/v3/repos/statuses/#get-the-combined-status-for-a-specific-ref) is used to retrieve statuses for the latest commit in an editorial workflow PR. The correct status for a deploy preview is inferred using a list of keywords, which currently consists of just "deploy", and selecting the first status whose context contains that keyword.

The inference behavior can be overridden by providing `preview_context` under the `backend` configuration. The provided value is expected to match the entire context string for a status to match.

### Link to the current entry
Generally a commit status will only provide a root URL, but editors clicking "Preview" when editing an entry will expect to see a preview of that entry. This requires configuration since Netlify CMS only knows where your content source lives in your repo, and not where it will end up on your site. They're usually not the same path.

Each collection in the config accepts a `preview_path` string. It may have up to one `{{slug}}` placeholder, which will be replaced with the entire slug (unlike when it is used in the `slug` configuration, confusingly - will clarify in the docs). The `preview_path` value is appended to the deploy preview URL retrieved from the backend.

Example:
```yaml
collection:
  name: blog
  slug: {{year}}-{{month}}-{{slug}}
  preview_path: blog/{{slug}} # `{{slug}}` here is replaced with the output of the entire "slug" configured in the line above
```

Closes #372 

## Remaining steps
- [x] Allow to be disabled (shoutout to @talves 😛)
- [x] Use `site_url` as default for `display_url`
- [x] Git Gateway support (https://github.com/netlify/git-gateway/issues/28)
- [x] Clear up newly introduced proptype warnings
- [x] Test with file collections
- [x] Test with unsupporting backends
- [x] Test with Git Gateway
- [x] Docs
- [x] Blog post
- [x] Custom slug support
- [x] Block `loadDeployPreview` when already fetching
- [x] Update config schema
- [x] Test on production site(s)

### CODE REVIEW

File: packages/netlify-cms-backend-github/src/implementation.js [hunk 1]
```diff
@@ -6,6 +6,34 @@ import API from './API';
 
 const MAX_CONCURRENT_DOWNLOADS = 10;
 
+/**
+ * Keywords for inferring a status that will provide a deploy preview URL.
+ */
+const PREVIEW_CONTEXT_KEYWORDS = ['deploy'];
```
Comment (tech4him1) [inline packages/netlify-cms-backend-github/src/implementation.js]: Should this be editable from the `config.yml`?

**Comment by indysigner** [comment]
Hi @erquhart , this is really great to see it happen! Our Smashing editors have been waiting for this for a loooong time! ;-)

So I just went over to the [demo CMS](https://deploy-preview-2028--cms-demo.netlify.com/#/workflow) and tried to preview a test article (see the below screenshot). But as soon as I'm saving the article, the preview link does actually go away (and doesn't come back either).

Can you please have a look, @erquhart ?

screenshot 2019-02-02 10 50 10

I also tried using the other Netlify CMS WWW preview, but it's asking me for write permissions to all of our & mine private repositories... 🤔 which is a _little_ scary.

**Comment by erquhart** [comment]
@indysigner hey there! The demo uses the test-repo, which doesn't support preview links. Once this is released it should work fine for them Smashing setup, we just need to get you migrated to 2.x first.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by indysigner** [comment]
Okie, thanks for letting me know, @erquhart! :-)

**Comment by sddaniels** [comment]
Awesome to see this coming along! At the risk of opening a small can of worms, I do have a question surrounding the `preview_path` laid out here:

As I was working on our implementation of Netlify CMS, one of the asks from our editors was to be able to create any number of pages and set whatever permalink they wanted for any particular page. Since we're migrating from GitHub Pages (no time for Gatsby yet, sadly), that led me to the `permalink` front matter property you can set in Jekyll. Letting editors enter a permalink value themselves fits the bill neatly, except it means that the slug in Netlify CMS doesn't correspond to the preview link anymore. I think that means the direct links won't work for us.

I might be missing something, in which case awesome. Otherwise I'm curious if there can be some sort of other option or replacement than `{{slug}}`, maybe one that is allowed to be user provided. I certainly don't want to hold up this PR since it's so close.

**Comment by erquhart** [comment]
That's exactly what's holding this PR up right now. I expect to have support for that when we release 👍🏼

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by sddaniels** [comment]
That's awesome, thanks @erquhart!

**Comment by indysigner** [comment]
Wow, exciting, thanks for your work! @erquhart

**Comment by erquhart** [comment]
No problem!
